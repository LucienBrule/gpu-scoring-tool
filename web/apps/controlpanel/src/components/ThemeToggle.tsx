'use client';

import React, { useEffect, useState } from 'react';

/**
 * ThemeToggle component for switching between light and dark mode
 * Used to test the Catppuccin Mocha theme implementation
 */
export function ThemeToggle() {
  const [theme, setTheme] = useState<'light' | 'dark'>('dark');

  // Initialize theme based on localStorage or system preference
  useEffect(() => {
    // Check if theme is stored in localStorage
    const storedTheme = localStorage.getItem('theme') as 'light' | 'dark' | null;
    
    if (storedTheme) {
      setTheme(storedTheme);
    } else if (window.matchMedia('(prefers-color-scheme: dark)').matches) {
      setTheme('dark');
    } else {
      setTheme('light');
    }
  }, []);

  // Update document class and localStorage when theme changes
  useEffect(() => {
    if (theme === 'dark') {
      document.documentElement.classList.add('dark');
      document.documentElement.classList.remove('light');
    } else {
      document.documentElement.classList.add('light');
      document.documentElement.classList.remove('dark');
    }
    
    localStorage.setItem('theme', theme);
  }, [theme]);

  // Toggle between light and dark mode
  const toggleTheme = () => {
    setTheme(theme === 'light' ? 'dark' : 'light');
  };

  return (
    <button
      onClick={toggleTheme}
      className="px-4 py-2 rounded-md bg-surface1 text-foreground hover:bg-surface2 transition-colors"
      aria-label={`Switch to ${theme === 'light' ? 'dark' : 'light'} mode`}
    >
      {theme === 'light' ? '🌙 Dark Mode' : '☀️ Light Mode'}
    </button>
  );
}

export default ThemeToggle;