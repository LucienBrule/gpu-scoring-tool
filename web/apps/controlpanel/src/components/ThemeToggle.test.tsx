import React from 'react';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen, act } from '@testing-library/react';
import ThemeToggle from './ThemeToggle';

// Mock localStorage
const localStorageMock = (() => {
  let store: Record<string, string> = {};
  return {
    getItem: vi.fn((key: string) => store[key] || null),
    setItem: vi.fn((key: string, value: string) => {
      store[key] = value.toString();
    }),
    clear: vi.fn(() => {
      store = {};
    }),
  };
})();

// Mock document methods
Object.defineProperty(window, 'localStorage', {
  value: localStorageMock,
});

// Mock matchMedia
Object.defineProperty(window, 'matchMedia', {
  value: vi.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: vi.fn(),
    removeListener: vi.fn(),
    addEventListener: vi.fn(),
    removeEventListener: vi.fn(),
    dispatchEvent: vi.fn(),
  })),
});

// Mock document.documentElement
const mockAdd = vi.fn();
const mockRemove = vi.fn();
document.documentElement.classList = {
  add: mockAdd,
  remove: mockRemove,
  // Add other required DOMTokenList methods as needed
  contains: vi.fn(),
  toggle: vi.fn(),
  replace: vi.fn(),
  supports: vi.fn(),
  value: '',
  length: 0,
  item: vi.fn(),
  forEach: vi.fn(),
  entries: vi.fn(),
  keys: vi.fn(),
  values: vi.fn(),
  [Symbol.iterator]: vi.fn(),
} as DOMTokenList;

describe('ThemeToggle Component', () => {
  beforeEach(() => {
    vi.clearAllMocks();
    localStorageMock.clear();
  });

  it('renders correctly in light mode', async () => {
    // Mock localStorage to return 'light'
    localStorageMock.getItem.mockReturnValueOnce('light');
    
    let container;
    await act(async () => {
      const result = render(<ThemeToggle />);
      container = result.container;
    });
    
    expect(container).toMatchSnapshot();
    
    // Check if the button text is correct for light mode
    expect(screen.getByText(/Dark Mode/i)).toBeInTheDocument();
  });

  it('renders correctly in dark mode', async () => {
    // Mock localStorage to return 'dark'
    localStorageMock.getItem.mockReturnValueOnce('dark');
    
    let container;
    await act(async () => {
      const result = render(<ThemeToggle />);
      container = result.container;
    });
    
    expect(container).toMatchSnapshot();
    
    // Check if the button text is correct for dark mode
    expect(screen.getByText(/Light Mode/i)).toBeInTheDocument();
  });

  it('defaults to dark mode when no preference is stored', async () => {
    // Mock localStorage to return null (no stored preference)
    localStorageMock.getItem.mockReturnValueOnce(null);
    
    // Mock matchMedia to return false for dark mode preference
    Object.defineProperty(window, 'matchMedia', {
      value: vi.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: vi.fn(),
        removeListener: vi.fn(),
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
        dispatchEvent: vi.fn(),
      })),
    });
    
    await act(async () => {
      render(<ThemeToggle />);
    });
    
    // Check if the button text is correct for light mode (default)
    expect(screen.getByText(/Dark Mode/i)).toBeInTheDocument();
  });

  it('uses system preference when no preference is stored', async () => {
    // Mock localStorage to return null (no stored preference)
    localStorageMock.getItem.mockReturnValueOnce(null);
    
    // Mock matchMedia to return true for dark mode preference
    Object.defineProperty(window, 'matchMedia', {
      value: vi.fn().mockImplementation((query) => ({
        matches: query === '(prefers-color-scheme: dark)',
        media: query,
        onchange: null,
        addListener: vi.fn(),
        removeListener: vi.fn(),
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
        dispatchEvent: vi.fn(),
      })),
    });
    
    await act(async () => {
      render(<ThemeToggle />);
    });
    
    // Check if the button text is correct for dark mode (system preference)
    expect(screen.getByText(/Light Mode/i)).toBeInTheDocument();
  });
});