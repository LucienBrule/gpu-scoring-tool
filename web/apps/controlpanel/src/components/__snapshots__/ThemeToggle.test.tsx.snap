// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ThemeToggle Component > renders correctly in dark mode 1`] = `
<div>
  <button
    aria-label="Switch to light mode"
    class="px-4 py-2 rounded-md bg-surface1 text-foreground hover:bg-surface2 transition-colors"
  >
    ☀️ Light Mode
  </button>
</div>
`;

exports[`ThemeToggle Component > renders correctly in light mode 1`] = `
<div>
  <button
    aria-label="Switch to dark mode"
    class="px-4 py-2 rounded-md bg-surface1 text-foreground hover:bg-surface2 transition-colors"
  >
    🌙 Dark Mode
  </button>
</div>
`;
